src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.InstallRecords.cs

```diff
@@ -0,0 +1,339 @@
+                    {
+                        Log.LogMessage($"Found apparent manifest package ID '{manifestPackageId} which did not correctly parse into manifest ID and feature band.");
+                        continue;
+                    }
+
+                    string manifestId = manifestPackageId.Substring(0, separatorIndex);
+                    string manifestFeatureBand = manifestPackageId.Substring(separatorIndex + ManifestSeparator.Length);
+
+                    using RegistryKey manifestKey = installedManifestsKey.OpenSubKey(manifestPackageId);
+                    foreach (string manifestVersion in manifestKey.GetSubKeyNames())
+                    {
+                        using RegistryKey manifestVersionKey = manifestKey.OpenSubKey(manifestVersion);
+
+                        WorkloadManifestRecord record = new WorkloadManifestRecord
+                        {
+                            ManifestId = manifestId,
+                            ManifestVersion = manifestVersion,
+                            ManifestFeatureBand = manifestFeatureBand,
+                            ProductCode = (string)manifestVersionKey.GetValue("ProductCode"),
+                            UpgradeCode = (string)manifestVersionKey.GetValue("UpgradeCode"),
+                            ProductVersion = new Version((string)manifestVersionKey.GetValue("ProductVersion")),
+                            ProviderKeyName = (string)manifestVersionKey.GetValue("DependencyProviderKey")
+                        };
+
+                        Log.LogMessage($"Found workload manifest record, Id: {manifestId}, version: {manifestVersion}, feature band: {manifestFeatureBand}, ProductCode: {record.ProductCode}, provider key: {record.ProviderKeyName}");
+                        manifestRecords.Add(record);
+                        discoveredManifests.Add((manifestId, manifestVersion));
+                    }
+                }
+            }
+
+            //  Workload manifest MSIs for 8.0.100 don't yet write the same type of installation records to the registry that workload packs do.
+            //  So to find what is installed, we look for the manifests on disk, and then map that to installed MSIs
+            //  To do the mapping to installed MSIs, we rely on the fact that the MSI UpgradeCode is generated in a stable fashion from the
+            //  NuGet package identity and platform: Utils.CreateUuid(UpgradeCodeNamespaceUuid, $"{Package.Identity};{Platform}")
+            //  The NuGet package identity used is the vanilla (non-MSI) manifest package, for example Microsoft.NET.Workload.Mono.ToolChain.Current.Manifest-8.0.100 version 8.0.0
+
+            //  Manifest IDs are lowercased on disk, we need to map them back to the original casing to generate the right UpgradeCode
+            Dictionary<string, string> manifestIDCasing = new Dictionary<string, string>();
+            foreach (var casedManifestID in new[]
```

Comment:
should we move the array to a static readonly private field and avoid allocating it each time the method runs?